File: DSA_python/DSA11027 ĐẾM SỐ NODE LÁ CÂY NHỊ PHÂN TÌM KIẾM CÂN BẰNG.py
class TreeNode:
    def __init__(self, val = 0, left = None, right = None):
        self.val = val
        self.left = left
        self.right = right

def add_to_tree(root, node):
    if root.val > node:
        if not root.left: root.left = TreeNode(node)
        else: add_to_tree(root.left, node)
    elif root.val < node:
        if not root.right: root.right = TreeNode(node)
        else: add_to_tree(root.right, node)

def change_to_tree(root, edges):
    if len(edges) == 0: return

    mid = (len(edges)-1)//2

    add_to_tree(root, edges[mid])

    if len(edges) > 1:
        left = edges[:mid]
        right = edges[mid+1:]

        change_to_tree(root, left)
        change_to_tree(root, right)

def find_leaves(node):
    leaves = list()

    if not node.left and not node.right:
        leaves.append(node.val)
        return leaves

    left, right = list(), list()

    if node.left: left = find_leaves(node.left)
    if node.right: right = find_leaves(node.right)

    leaves = left[:] + right[:]
    return leaves

def submit():
    n = int(input())
    edges = list(map(int, input().split()))
    edges.sort()

    root = TreeNode(edges[(n-1)//2])
    change_to_tree(root, edges)

    leaves = find_leaves(root)

    print(len(leaves))











test = int(input())
for i in range(test):
    submit()

File: DSA/DSA11027 ĐẾM SỐ NODE LÁ CÂY NHỊ PHÂN TÌM KIẾM CÂN BẰNG.cpp
#include <bits/stdc++.h>
using namespace std;

struct TreeNode
{
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int v) : val(v), left(nullptr), right(nullptr) {}
};


void add(TreeNode** root, int node)
{
    if (!(*root))
    {
        (*root) = new TreeNode(node);
    }
    else if (node < (*root)->val)
    {
        if (!(*root)->left) (*root)->left = new TreeNode(node);
        else add(&(*root)->left, node);
    }
    else if (node > (*root)->val)
    {
        if (!(*root)->right) (*root)->right = new TreeNode(node);
        else add(&(*root)->right, node);
    }
}

void insert(TreeNode** root, vector<int> node)
{
    if (node.empty())
    {
        return;
    }
    
    int mid = (node.size()-1)/2;

    add(&(*root), node[mid]);

    if (node.size() > 1)
    {
        vector<int> left(node.begin(), node.begin()+mid);
        vector<int> right(node.begin()+mid+1, node.end());
        
        insert(root, left);
        insert(root, right);
    }
}

vector<int> find_leaves(TreeNode* node)
{
    vector<int> leaves;

    if (!node->left && !node->right)
    {
        leaves.push_back(node->val);
        return leaves;
    }
    
    vector<int> left, right;
    if (node->left) left = find_leaves(node->left);
    if (node->right) right = find_leaves(node->right);

    leaves.insert(leaves.end(), left.begin(), left.end());
    leaves.insert(leaves.end(), right.begin(), right.end());

    return leaves;
}

void submit()
{
    int n;
    cin >> n;
    vector<int> node(n);
    for (auto &&i : node) cin >> i;

    sort(node.begin(), node.end());

    TreeNode* root = nullptr;

    insert(&root, node);

    vector<int> leaves = find_leaves(root);

    cout << leaves.size();

    cout << endl;
}

int main()
{
    ios::sync_with_stdio(false);                                    // Speed up input/output
    cin.tie(nullptr);
    cout.tie(nullptr);

    int so_bo_test;
    cin >> so_bo_test;
    cin.ignore();
    
    while (so_bo_test--)
    {
        submit();
    }

    return 0;
} 